runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkProcessContext.java

```diff
@@ -137,31 +131,43 @@ protected OutputT computeNext() {
       // collection (and iterator) is reset between each call to processElement, so the
       // collection only holds the output values for each call to processElement, rather
       // than for the whole partition (which would use too much memory).
-      while (true) {
-        if (outputIterator.hasNext()) {
-          return outputIterator.next();
-        } else if (inputIterator.hasNext()) {
-          clearOutput();
-          // grab the next element and process it.
-          doFnRunner.processElement(inputIterator.next());
-          outputIterator = getOutputIterator();
-        } else if (timerDataIterator.hasNext()) {
+      try {
+        boolean isBundleStarted = false;
+        boolean isBundleFinished = false;
+        while (true) {
+          if (outputIterator.hasNext()) {
+            return outputIterator.next();
+          }
+
+          if (!isBundleStarted) {
+            isBundleStarted = true;
+            // call startBundle() before beginning to process the partition.
+            doFnRunner.startBundle();
```

Comment:
Is there a reason not to call `doFnRunner.startBundle()` before entering the `while` loop? at this point in the code it should be called, right? and it would save the whole "if/flag" thing..

Reply:
The `doFnRunner.startBundle()` call should go inside of the computeNext iterator so we cover the case of calling teardown on an exception on startBundle (this is not covered in the current version so we are solving an extra bug). It is called inside of the while loop mostly for readability you have the complete start-process-finish bundle lifecycle in one place (I know that isBundleStarted is far from beautiful but at least we have the full state machine in one place). Also the logic of are we outputting (the first if) is separated from are we processing a bundle and this is now explicit.

Reply:
call startBundle inside the `try` before the `while` should work too, but that's really more about the codestyle and just a nit.